src/layer.cpp

```diff
@@ -214,17 +223,114 @@ scene_on_touch_begin(VALUE rcv, SEL sel)
     }
     block = rb_retain(block); // FIXME need release...
 
-    auto listener = cocos2d::EventListenerTouchOneByOne::create();
-    listener->onTouchBegan = [block](cocos2d::Touch *touch,
+    auto scene = SCENE(rcv);
+    if (!RTEST(scene->touch_listener)) {
+        scene->touch_listener = cocos2d::EventListenerTouchOneByOne::create();
+    } else {
```

Comment:
Could you replace this with:

```
}
else {
```

to conform to the project's coding style :-)